### PR TITLE
adding pre-commit hooks

### PR DESCRIPTION
**Pull request by axiomcura**
This small PR introduces `pre-commit`s for `cytosnake`. 

The main file to look at is the `.pre-commit-config.yaml` file, which contains the programs that will be executed in order to keep the `cytosnake`'s code base to a specific standard.

### CODE REVIEW

File: docs/.pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+---
+repos:
+  # import formatter with black configurations
+    - repo: https://github.com/pycqa/isort
+      rev: 5.10.1
```
Comment (d33bs) [inline docs/.pre-commit-config.yaml:5]: I noticed a number of these dependencies might reference non-current revisions, for example this one. Consider using `pre-commit autoupdate` to automatically update these to the latest.
```suggestion
      rev: 5.11.4
```

File: docs/.pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      hooks:
+          - id: isort
+            name: isort (python)
+            args: [--profile, black, --filter-files]
+
+  # Code formatter for both python files and jupyter notebooks
+    - repo: https://github.com/psf/black
+      rev: 22.10.0
+      hooks:
+          - id: black-jupyter
```
Comment (d33bs) [inline docs/.pre-commit-config.yaml:15]: Love this! 🙌

File: docs/.pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    # snakemake formatting
+    - repo: https://github.com/snakemake/snakefmt
+      rev: v0.8.0
+      hooks:
+          - id: snakefmt
+
+    # fast linter written in rust
+    - repo: https://github.com/charliermarsh/ruff-pre-commit
+    # Ruff version.
+      rev: v0.0.218
```
Comment (d33bs) [inline docs/.pre-commit-config.yaml:28]: Consider revising the comment to match the indent of the line below.
```suggestion
      # Ruff version.
      rev: v0.0.218
```

File: docs/.pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      hooks:
+          - id: ruff
+        # Respect `exclude` and `extend-exclude` settings.
+            args: [--force-exclude]
```
Comment (d33bs) [inline docs/.pre-commit-config.yaml:32]: Consider revising the comment to match the indentation of the line below.
```suggestion
            # Respect `exclude` and `extend-exclude` settings.
            args: [--force-exclude]
```

File: docs/.pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,46 @@
+---
+repos:
+  # import formatter with black configurations
+    - repo: https://github.com/pycqa/isort
```
Comment (d33bs) [inline docs/.pre-commit-config.yaml:4]: Consider revising the comment to match the indentation of the line below.
```suggestion
    # import formatter with black configurations
    - repo: https://github.com/pycqa/isort
```

File: docs/.pre-commit-config.yaml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      rev: 5.10.1
+      hooks:
+          - id: isort
+            name: isort (python)
+            args: [--profile, black, --filter-files]
+
+  # Code formatter for both python files and jupyter notebooks
+    - repo: https://github.com/psf/black
```
Comment (d33bs) [inline docs/.pre-commit-config.yaml:12]: Consider revising the comment to match the indentation of the line below.
```suggestion
    # Code formatter for both python files and jupyter notebooks
    - repo: https://github.com/psf/black
```

**Review by d33bs — COMMENTED** [review]
Great stuff! Excited to see the pre-commit checks you implemented. I added a few thoughts with this review and also had the following additional questions for your consideration:

- I noticed that `.pre-commit-config.yaml` is found under the `docs` dir. Would it make sense to move it to the root or a root `.config` directory?
- Would it make sense to add developer or contributor documentation outlining how one might use pre-commit with this repository?
- While it's deprecated (reference with caution), there is a [Github Action for using pre-commit to automatically run checks](https://github.com/pre-commit/action) which might inspire future work towards continuous automated and observable testing for this repo. I mention this as optional / informational and not requirement from the review.

**Review by roshankern — APPROVED** [review]
Besides some indentation inconsistencies, very clean PR! Merge at will!

**Comment by axiomcura** [comment]
Formatting issues are all fixed! 

Comments from @d33bs 

> I noticed that .pre-commit-config.yaml is found under the docs dir. Would it make sense to move it to the root or a root .config directory?

That was a mistake! Moved it to the root directory folder. 

> Would it make sense to add developer or contributor documentation outlining how one might use pre-commit with this repository?

Yes, this is a good point. However, my assumption of pre-commits is that it is automatically executed when a developer submits a commit. Are you suggesting documenting the types of tools that are executed in the pre-commit configs? 

In addition, `pre-commit` does provide developers to manually execute the `pre-commit` hooks by typing `pre-commit run`.

I feel that this can be written in the README and in the sphinx docs.

**Comment by d33bs** [comment]
@axiomcura - Thanks for the replies, and nice work making changes! Replying to your question below.

> Yes, this is a good point. However, my assumption of pre-commits is that it is automatically executed when a developer submits a commit. Are you suggesting documenting the types of tools that are executed in the pre-commit configs?

Tools like pre-commit can help reduce the time and development friction by turning assumptions into reproducible expectations for a repository. However, because pre-commit may be used in so many ways (as a hook, as a manual check, within automated tooling) and there are fewer guarantees about a developer's environment + workflow it might be helpful to consider spelling out how you see it being used. For example, in some repos it's a suggestion for development, in others it may be a requirement. From the perspective of this repo, how do you see it being used by yourself or others (this could go into the docs)?

**Comment by axiomcura** [comment]
@d33bs Ahh I see what you're saying. 

Since I am using a package called `pre-commit` I should let users know on how to use it. You are right, there are many methods on how to use `pre-commits` I just happen to use it by downloading a package known as `pre-commit`; therefore, I must explicitly state on how the `pre-commits` are used through this tool. 

Correct me if I am wrong.

**Comment by d33bs** [comment]
@axiomcura - thanks, yes, I recommend being clear within documentation on how pre-commit should be used. For example, if another developer were to make a contribution to this repo, how likely would it be that their work would require corrections which could have otherwise been caught by pre-commit (and not a reviewer)? If they could have used pre-commit in such a way to prepare the code in advance of a review, I'd recommend documenting the suggested or required methods for doing so. Specifically I'm meaning running `pre-commit` for this repo in a broad sense, and not which checks or how those checks work (which is outlined within the configuration file).

**Comment by axiomcura** [comment]
@d33bs Thanks Dave. I'll make this an issue and take care of once we are close to production! Thank you for your insights!

**Pull request merged**